src/Crisp/Parser.hs

```diff
@@ -0,0 +1,188 @@
+module Crisp.Parser
+  ( atom
+  , sexpr
+  , crisp
+  ) where
+
+import Data.Foldable (asum)
+
+import Crisp.Datatypes
+import ParserCombinators.Datatypes
+import ParserCombinators.Parser
+
+-----------------------
+---- Crisp Related ----
+-----------------------
+
+-- | Parse an atom
+atom :: Parser Char [] String Atom
+atom =
+  asum
+    [ AInteger <$> integer
+    , ABool <$> bool
+    , ASymbol <$> symbol
+    ]
+
+-- | Parse a S expression
+sexpr :: Parser Char [] String [Crisp]
+sexpr =
+  asum $
+    sexprWith
+      <$> [ ("(", ")")
+          , ("[", "]")
+          , ("{", "}")
+          , ("<", ">")
+          ]
+  where
+    sexprWith (left, right) =
+      between
+        (string left)
+        (string right)
+        (crisp `sepBy1` spaces)
+
+-- -- | Parse a function definition
+-- function :: Parser Char [] String Function
+-- function = do
+--   _ <- string "("
+--   spaces
+--   _ <- string "fn"
+--   spaces
+--   name <- try symbol
+--   spaces
+--   args <- between (string "[") (string "]") (symbol `sepBy` spaces)
+--   spaces
+--   body <- crisp `sepBy` spaces
+--   spaces
```

Comment:
As you've seen here, and elsewhere, you end up putting a lot of `spaces` everywhere.

This can be improved by taking a more principled approach/convention regarding tokens and whitespace - I remember it being well described here - https://markkarpov.com/tutorial/megaparsec.html#lexing